File: java-to-go-projects/16-cron-parser/NextTime.java
package cron;

import java.time.*;
import java.util.*;

/**
 * Calculates the next N execution times for a given Schedule from a start time.
 */
public class NextTime {

    /**
     * Find the next N execution times starting from 'from'.
     */
    public static List<LocalDateTime> nextN(Schedule schedule, LocalDateTime from, int count) {
        List<LocalDateTime> results = new ArrayList<>();
        LocalDateTime current = from.plusMinutes(1).withSecond(0).withNano(0); // start from next minute

        while (results.size() < count) {
            if (matches(schedule, current)) {
                results.add(current);
            }
            current = current.plusMinutes(1);

            // Safety: don't loop forever (max 2 years ahead)
            if (current.isAfter(from.plusYears(2))) {
                break;
            }
        }

        return results;
    }

    /**
     * Check if a given time matches the schedule.
     */
    public static boolean matches(Schedule schedule, LocalDateTime time) {
        return schedule.getMinutes().contains(time.getMinute())
                && schedule.getHours().contains(time.getHour())
                && schedule.getDaysOfMonth().contains(time.getDayOfMonth())
                && schedule.getMonths().contains(time.getMonthValue())
                && schedule.getDaysOfWeek().contains(time.getDayOfWeek().getValue() % 7);
    }
}

File: java-to-go-projects/16-cron-parser/rust/src/next_time.rs
use crate::schedule::Schedule;

/// Check if a given date/time matches the schedule.
pub fn matches(
    schedule: &Schedule,
    minute: u32,
    hour: u32,
    day: u32,
    month: u32,
    weekday: u32,
) -> bool {
    todo!()
}

/// Calculate the next N execution times from a starting point.
/// Returns Vec of (year, month, day, hour, minute) tuples.
/// Iterates minute-by-minute from start.
pub fn next_n(
    schedule: &Schedule,
    start_year: i32,
    start_month: u32,
    start_day: u32,
    start_hour: u32,
    start_minute: u32,
    count: usize,
) -> Vec<(i32, u32, u32, u32, u32)> {
    todo!()
}
